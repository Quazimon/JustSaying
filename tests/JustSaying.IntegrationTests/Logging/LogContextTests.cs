using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using JustSaying.IntegrationTests.Fluent;
using JustSaying.Messaging;
using JustSaying.Messaging.MessageHandling;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Shouldly;
using Xunit.Abstractions;

namespace JustSaying.Logging
{
    public class LogContextTests : IntegrationTestBase
    {
        public LogContextTests(ITestOutputHelper outputHelper) : base(outputHelper)
        { }

        [AwsFact]
        public async Task PublishToTopicLogsShouldHaveContext()
        {
            var services = GivenJustSaying()
                .ConfigureJustSaying(
                    (builder) => builder.WithLoopbackTopic<SimpleMessage>(UniqueName));

            var sp = services.BuildServiceProvider();

            var cts = new CancellationTokenSource();

            var publisher = sp.GetRequiredService<IMessagePublisher>();
            await publisher.StartAsync(cts.Token);

            var message = new SimpleMessage();
            await publisher.PublishAsync(message, cts.Token);

            var output = OutputHelper.Output;
            output.ShouldMatchApproved(o => o
                .SubFolder("Approvals")
                .WithScrubber(logMessage => ScrubLogs(logMessage, message.Id.ToString())));

            cts.Cancel();
        }

        [AwsFact]
        public async Task PublishToQueueLogsShouldHaveContext()
        {
            var services = GivenJustSaying()
                .ConfigureJustSaying(
                    (builder) => builder.WithLoopbackQueue<SimpleMessage>(UniqueName));

            var sp = services.BuildServiceProvider();

            var cts = new CancellationTokenSource();

            var publisher = sp.GetRequiredService<IMessagePublisher>();
            await publisher.StartAsync(cts.Token);

            var message = new SimpleMessage();
            await publisher.PublishAsync(message, cts.Token);

            var output = OutputHelper.Output;
            output.ShouldMatchApproved(o => o
                .SubFolder("Approvals")
                .WithScrubber(logMessage => ScrubLogs(logMessage, message.Id.ToString())));

            cts.Cancel();
        }

        [AwsFact]
        public async Task HandleMessageFromQueueLogs_ShouldHaveContext()
        {
            var handler = new InspectableHandler<SimpleMessage>();

            var services = GivenJustSaying()
                .ConfigureJustSaying(
                    (builder) => builder.WithLoopbackQueue<SimpleMessage>(UniqueName))
                .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);

            var sp = services.BuildServiceProvider();

            var cts = new CancellationTokenSource();

            var publisher = sp.GetRequiredService<IMessagePublisher>();
            await publisher.StartAsync(cts.Token);
            await sp.GetRequiredService<IMessagingBus>().StartAsync(cts.Token);

            var message = new SimpleMessage();
            await publisher .PublishAsync(message, cts.Token);

            await Patiently.AssertThatAsync(OutputHelper,
                () =>
                {
                    handler.ReceivedMessages
                        .ShouldHaveSingleItem()
                        .Id.ShouldBe(message.Id);
                });

            var output = ((TestOutputHelper) OutputHelper).Output;
            output.ShouldMatchApproved(o => o
                .SubFolder("Approvals")
                .WithScrubber(logMessage => ScrubLogs(logMessage, message.Id.ToString())));

            cts.Cancel();
        }

        private string ScrubLogs(string message, string messageId)
        {
            message = message.Replace(messageId, "{MessageId}");
            message = message.Replace(UniqueName, "{TestDiscriminator}");

            message = Regex.Replace(message, @"AwsRequestId: .{8}-.{4}-.{4}-.{4}-.{12}", "AwsRequestId: {AwsRequestId}");
            message = Regex.Replace(message, @"(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})Z", "{DateTime}");
            message = Regex.Replace(message, @"(\d+).(\d+) ms", "{Duration}");
            message = Regex.Replace(message, @"in (\d+)ms.", "{Duration}");
            return message;
        }
    }
}
