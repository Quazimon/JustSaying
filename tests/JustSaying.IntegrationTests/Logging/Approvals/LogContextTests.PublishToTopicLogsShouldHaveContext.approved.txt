[{DateTime}] info: JustSaying.Fluent.TopicPublicationBuilder[0]
      Adding SNS publisher for message type 'JustSaying.TestingFramework.SimpleMessage'.
[{DateTime}] info: JustSaying.Fluent.TopicPublicationBuilder[0]
      Created SNS topic publisher on topic 'simplemessage' for message type 'JustSaying.TestingFramework.SimpleMessage'.
[{DateTime}] dbug: JustSaying[0]
      Created topic 'simplemessage' with ARN 'arn:aws:sns:us-east-1:000000000000:simplemessage'.
[{DateTime}] info: JustSaying[0]
      Running 1 startup tasks
[{DateTime}] info: JustSaying[0]
      Starting bus with settings: JustSaying.Messaging.Interrogation.InterrogationResult
[{DateTime}] info: JustSaying.Messaging.Channels.SubscriptionGroups.SubscriptionGroupCollection[0]
      Subscription group collection successfully started
[{DateTime}] info: JustSaying[0]
      Starting bus
[{DateTime}] info: JustSaying.Publish[0]
      => AwsRequestId: {AwsRequestId}
      Published message {MessageId} of type SimpleMessage to Topic 'arn:aws:sns:us-east-1:000000000000:simplemessage'.
